6238_prototype/src/main/java/frc/robot/commands/ShooterCommand.java

```diff
@@ -5,22 +5,67 @@
 import frc.robot.subsystems.BallSubsystem;
 
 public class ShooterCommand extends CommandBase{
+    public static enum PneumaticKickers {LEFT_KICKER, RIGHT_KICKER, BOTH};
+
+    private final PneumaticKickers kicker;
     private final BallSubsystem ball; 
-    private final SmartDashboardParam upperShooterSpeedTarget;
-    public ShooterCommand(BallSubsystem ball) {
+    private final double upperShooterRPMTarget;
+    private long timerStart;
+    private final long timeLimit;
+    private final double rpmDiffTolerance;
+    private boolean pneumaticsActivated;
+    private long countdownStart;
+
+    public ShooterCommand(BallSubsystem ball, PneumaticKickers kicker, double rpmTarget) {
         this.ball = ball;
-        upperShooterSpeedTarget = new SmartDashboardParam("upperMotorSpeedTarget");
+        upperShooterRPMTarget = rpmTarget;
+        timeLimit = 2000;
+        timerStart = System.currentTimeMillis();
+        countdownStart = Long.MAX_VALUE;
+        rpmDiffTolerance = 50;
+        this.kicker = kicker;
+        pneumaticsActivated = true;
         addRequirements(ball);
     }
 
+    private void restartTimer() {
+        timerStart = System.currentTimeMillis();
+    }
+
     @Override
     public void execute() {
-        ball.retract();
-        ball.setSpeed(0.0, upperShooterSpeedTarget.get());
+        if (ball.getRPMUpperMotor() < upperShooterRPMTarget - rpmDiffTolerance
+            || upperShooterRPMTarget + rpmDiffTolerance < ball.getRPMUpperMotor()) {
+            restartTimer();
```

Comment:
I think this would be more clear if you used `Math.abs`:
```
if (Math.abs(ball.getRPMUpperMotor() - upperShooterRPMTarget) > rpmDiffTolerance) {
  restartTimer();
```